loguru/_handler.py

```diff
@@ -95,6 +98,24 @@ def __init__(
     def __repr__(self):
         return "(id=%d, level=%d, sink=%s)" % (self._id, self._levelno, self._name)
 
+    @contextmanager
+    def _protected_lock(self):
+        """Acquire the lock, but fail fast if its already acquired by the current thread."""
+        if getattr(self._lock_acquired, "acquired", False):
+            raise RuntimeError(
+                "Could not acquire internal lock because it was already in use (deadlock avoided). "
+                "This likely happened because the logger was re-used inside a sink, a signal "
+                "handler or a '__del__' method. This is not permitted because the logger and its "
+                "handlers are not re-entrant."
+            )
+        self._lock_acquired.acquired = True
+        try:
```

Comment:
I would replace this block with:

```python
self._lock_acquired.acquired = True
with self._lock:
    yield
self._lock_acquired.acquired = False
```

Is theer any reason to use `try / finally`?

Reply:
I think so. If an exception is raised, then the lock would be released, but the `selc._lock_acquired.acquired` would still be `True`.

Try this fake example:

```python
from contextlib import contextmanager

@contextmanager
def fake_lock():
    try:
        print("lock acquired")
        yield
    finally:
        print("lock release")


@contextmanager
def fake_protected_lock():
    print("lock state set to True")
    with fake_lock():
        yield
    print("lock state set to False")


with fake_protected_lock():
    raise ValueError

```

Reply:
My bad, I don't know why I assumed it was automatically handled, I'm probably too used to `pytest` fixtures. The `try / finally` is required as explained in the docs I shared myself... :man_facepalming: 

Thanks for the example. :+1: